File: agent/python/notion.py
"""Notion is the live work ledger. Channels stay the markdown wall.

The agent reads the Projects database and writes back:
  - Broadcast TS + Broadcast Permalink after a post
  - Last Status Update after a gather (the breadcrumb trail)

Stdlib HTTP only. Token from NOTION_TOKEN. Dry-run queries and prints
writes; it does not PATCH.
"""

from __future__ import annotations

import json
import os
from typing import Any, Callable
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from mapping import Mapping, Work

API = "https://api.notion.com/v1"
NOTION_VERSION = "2022-06-28"
MAX_BLOCK = 2000

PROPS = {
    "name": "Name",
    "id": "Scope key",
    "room": "Admin Channel",
    "state": "Status",
    "anchor": "Broadcast TS",
    "owner": "Owner",
    "permalink": "Broadcast Permalink",
    "status_update": "Last Status Update",
}

STATES = {
    "active": "active",
    "aligning": "active",
    "in progress": "active",
    "in-progress": "active",
    "blocked": "blocked",
    "closed": "closed",
    "done": "closed",
    "parked": "closed",
}

Transport = Callable[[str, str, dict[str, str], bytes | None], dict[str, Any]]


class NotionError(RuntimeError):
    pass


def default_transport(method: str, url: str, headers: dict[str, str], body: bytes | None) -> dict[str, Any]:
    req = Request(url, data=body, method=method, headers=headers)
    try:
        with urlopen(req, timeout=20) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except HTTPError as err:
        raise NotionError(f"notion {method} {err.code}") from err
    except URLError as err:
        raise NotionError("notion unreachable") from err


class NotionClient:
    def __init__(self, token: str, transport: Transport | None = None) -> None:
        if not token:
            raise NotionError("NOTION_TOKEN is required")
        self.token = token
        self.transport = transport or default_transport

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": "Bearer " + self.token,
            "Notion-Version": NOTION_VERSION,
            "Content-Type": "application/json",
            "User-Agent": "axiom-relay/0.3",
        }

    def _call(self, method: str, path: str, payload: dict[str, Any] | None = None) -> dict[str, Any]:
        body = json.dumps(payload).encode("utf-8") if payload is not None else None
        return self.transport(method, API + path, self._headers(), body)

    def query_work(self, database_id: str) -> tuple[Work, ...]:
        rows: list[Work] = []
        cursor: str | None = None
        while True:
            payload: dict[str, Any] = {"page_size": 100}
            if cursor:
                payload["start_cursor"] = cursor
            data = self._call("POST", "/databases/" + database_id + "/query", payload)
            for page in data.get("results", []):
                w = work_from_page(page)
                if w:
                    rows.append(w)
            if not data.get("has_more"):
                break
            cursor = data.get("next_cursor")
        return tuple(rows)

    def write_anchor(self, page_id: str, ts: str, permalink: str) -> None:
        self._call(
            "PATCH",
            "/pages/" + page_id,
            {
                "properties": {
                    PROPS["anchor"]: rich_text(ts),
                    PROPS["permalink"]: rich_text(permalink),
                }
            },
        )

    def write_trail(self, page_id: str, trail: str) -> None:
        self._call(
            "PATCH",
            "/pages/" + page_id,
            {"properties": {PROPS["status_update"]: rich_text(clip(trail))}},
        )


def clip(text: str) -> str:
    if len(text) <= MAX_BLOCK:
        return text
    suffix = "\n(truncated)"
    return text[: MAX_BLOCK - len(suffix)] + suffix


def rich_text(content: str) -> dict[str, Any]:
    return {"rich_text": [{"type": "text", "text": {"content": content}}]}


def _plain(prop: dict[str, Any] | None) -> str:
    if not prop:
        return ""
    kind = prop.get("type")
    if kind == "title":
        return "".join(t.get("plain_text", "") for t in prop.get("title", []))
    if kind == "rich_text":
        return "".join(t.get("plain_text", "") for t in prop.get("rich_text", []))
    if kind in {"select", "status"}:
        node = prop.get(kind) or {}
        return str(node.get("name") or "")
    if kind == "url":
        return str(prop.get("url") or "")
    if kind == "number" and prop.get("number") is not None:
        return str(prop["number"])
    return ""


def normalize_state(raw: str) -> str:
    return STATES.get(raw.strip().lower(), "active" if raw else "active")


def work_from_page(page: dict[str, Any]) -> Work | None:
    props = page.get("properties") or {}
    name = _plain(props.get(PROPS["name"]))
    key = _plain(props.get(PROPS["id"])) or name.lower().replace(" ", "-")
    if not key:
        return None
    return Work(
        id=key,
        name=name or key,
        room=_plain(props.get(PROPS["room"])),
        state=normalize_state(_plain(props.get(PROPS["state"]))),
        anchor=_plain(props.get(PROPS["anchor"])),
        owner=_plain(props.get(PROPS["owner"])),
        page_id=str(page.get("id") or ""),
    )


def load_work_from_notion(mapping: Mapping, database_id: str, client: NotionClient) -> Mapping:
    return mapping.replace_work(client.query_work(database_id))


def apply_writes(
    client: NotionClient,
    plan: tuple[tuple[Work, str], ...],
    trails: dict[str, str],
    *,
    dry_run: bool,
) -> list[str]:
    """Write anchors and trails back to Notion. Dry-run returns the plan, no PATCH."""
    lines: list[str] = []
    for w, action in plan:
        if not w.page_id:
            lines.append(f"notion  {w.id:16}  skip  no page id")
            continue
        if action == "post":
            lines.append(f"notion  {w.id:16}  POST    write Broadcast TS + Permalink")
            if not dry_run:
                client.write_anchor(w.page_id, w.anchor, trails.get(w.id, ""))
        elif action == "gather":
            lines.append(f"notion  {w.id:16}  GATHER  write Last Status Update")
            if not dry_run:
                client.write_trail(w.page_id, trails.get(w.id, "silent"))
        elif action == "gap":
            lines.append(f"notion  {w.id:16}  GAP     no write")
        else:
            lines.append(f"notion  {w.id:16}  SKIP    closed")
    return lines


def client_from_env(transport: Transport | None = None) -> NotionClient:
    return NotionClient(os.environ.get("NOTION_TOKEN", ""), transport)

File: agent/node/notion.ts
/** Notion is the live work ledger. Channels stay the markdown wall.
 *
 * Reads the Projects database and writes Broadcast TS, Broadcast Permalink,
 * and Last Status Update. Stdlib fetch. Token from NOTION_TOKEN.
 * Dry-run queries and prints writes; it does not PATCH.
 */
import type { Mapping, Work } from "./mapping.ts";
import { replaceWork } from "./mapping.ts";

const API = "https://api.notion.com/v1";
const NOTION_VERSION = "2022-06-28";
const MAX_BLOCK = 2000;

export const PROPS = {
  name: "Name",
  id: "Scope key",
  room: "Admin Channel",
  state: "Status",
  anchor: "Broadcast TS",
  owner: "Owner",
  permalink: "Broadcast Permalink",
  statusUpdate: "Last Status Update",
} as const;

const STATES: Record<string, string> = {
  active: "active",
  aligning: "active",
  "in progress": "active",
  "in-progress": "active",
  blocked: "blocked",
  closed: "closed",
  done: "closed",
  parked: "closed",
};

export type Transport = (method: string, url: string, headers: Record<string, string>, body?: string) => Promise<any>;

export class NotionError extends Error {}

async function defaultTransport(method: string, url: string, headers: Record<string, string>, body?: string) {
  const res = await fetch(url, { method, headers, body });
  if (!res.ok) throw new NotionError(`notion ${method} ${res.status}`);
  return await res.json();
}

export class NotionClient {
  constructor(
    private token: string,
    private transport: Transport = defaultTransport,
  ) {
    if (!token) throw new NotionError("NOTION_TOKEN is required");
  }

  private headers() {
    return {
      Authorization: `Bearer ${this.token}`,
      "Notion-Version": NOTION_VERSION,
      "Content-Type": "application/json",
      "User-Agent": "axiom-relay/0.3",
    };
  }

  private call(method: string, path: string, payload?: unknown) {
    return this.transport(method, API + path, this.headers(), payload ? JSON.stringify(payload) : undefined);
  }

  async queryWork(databaseId: string): Promise<Work[]> {
    const rows: Work[] = [];
    let cursor: string | undefined;
    while (true) {
      const payload: Record<string, unknown> = { page_size: 100 };
      if (cursor) payload.start_cursor = cursor;
      const data = await this.call("POST", `/databases/${databaseId}/query`, payload);
      for (const page of data.results ?? []) {
        const w = workFromPage(page);
        if (w) rows.push(w);
      }
      if (!data.has_more) break;
      cursor = data.next_cursor;
    }
    return rows;
  }

  writeAnchor(pageId: string, ts: string, permalink: string) {
    if (!ts) throw new NotionError("refusing empty Broadcast TS");
    return this.call("PATCH", `/pages/${pageId}`, {
      properties: {
        [PROPS.anchor]: richText(ts),
        [PROPS.permalink]: richText(permalink),
      },
    });
  }

  writeTrail(pageId: string, trail: string) {
    return this.call("PATCH", `/pages/${pageId}`, {
      properties: { [PROPS.statusUpdate]: richText(clip(trail)) },
    });
  }
}

export function clip(text: string): string {
  if (text.length <= MAX_BLOCK) return text;
  const suffix = "\n(truncated)";
  return text.slice(0, MAX_BLOCK - suffix.length) + suffix;
}

function richText(content: string) {
  return { rich_text: [{ type: "text", text: { content } }] };
}

function plain(prop: any): string {
  if (!prop) return "";
  if (prop.type === "title") return (prop.title ?? []).map((t: any) => t.plain_text ?? "").join("");
  if (prop.type === "rich_text") return (prop.rich_text ?? []).map((t: any) => t.plain_text ?? "").join("");
  if (prop.type === "select" || prop.type === "status") return prop[prop.type]?.name ?? "";
  if (prop.type === "url") return prop.url ?? "";
  if (prop.type === "number" && prop.number != null) return String(prop.number);
  return "";
}

export function workFromPage(page: any): Work | null {
  const props = page.properties ?? {};
  const name = plain(props[PROPS.name]);
  const key = plain(props[PROPS.id]) || name.toLowerCase().replace(/ /g, "-");
  if (!key) return null;
  const raw = plain(props[PROPS.state]).trim().toLowerCase();
  return {
    id: key,
    name: name || key,
    room: plain(props[PROPS.room]),
    state: STATES[raw] ?? "active",
    anchor: plain(props[PROPS.anchor]),
    owner: plain(props[PROPS.owner]),
    pageId: page.id ?? "",
  };
}

export async function loadWorkFromNotion(mapping: Mapping, databaseId: string, client: NotionClient): Promise<Mapping> {
  return replaceWork(mapping, await client.queryWork(databaseId));
}

export async function applyWrites(
  client: NotionClient,
  plan: { work: Work; action: string }[],
  posts: Record<string, { ts: string; permalink: string }>,
  trails: Record<string, string>,
  dryRun: boolean,
): Promise<string[]> {
  const lines: string[] = [];
  for (const { work: w, action } of plan) {
    if (!w.pageId) {
      lines.push(`notion  ${w.id.padEnd(16)}  skip  no page id`);
      continue;
    }
    if (action === "post") {
      const posted = posts[w.id];
      if (!posted?.ts) {
        lines.push(`notion  ${w.id.padEnd(16)}  POST    no ts — leave empty, retry next run`);
        continue;
      }
      lines.push(`notion  ${w.id.padEnd(16)}  POST    write Broadcast TS + Permalink`);
      if (!dryRun) await client.writeAnchor(w.pageId, posted.ts, posted.permalink);
    } else if (action === "gather") {
      lines.push(`notion  ${w.id.padEnd(16)}  GATHER  write Last Status Update`);
      if (!dryRun) await client.writeTrail(w.pageId, trails[w.id] ?? "silent");
    } else if (action === "gap") {
      lines.push(`notion  ${w.id.padEnd(16)}  GAP     no write`);
    } else {
      lines.push(`notion  ${w.id.padEnd(16)}  SKIP    closed`);
    }
  }
  return lines;
}

export function clientFromEnv(transport?: Transport): NotionClient {
  return new NotionClient(process.env.NOTION_TOKEN ?? "", transport);
}
